### PR TITLE
Budget entity, BudgetDaoImpl and tests

### PR DESCRIPTION
**Pull request by strutinsky**

### CODE REVIEW

File: src/main/java/com/smartcity/dao/BudgetDao.java [hunk 1]
```diff
@@ -8,10 +8,10 @@
 
     Budget create(Budget budget);
 
-    Budget get(long id);
```
Comment (seregamazur) [inline src/main/java/com/smartcity/dao/BudgetDao.java]: It must be Budget get(Long id),delete(Long id).

File: src/test/java/com/smartcity/dao/BudgetDaoImplTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Test
+    public void testDeleteBudget_noRowsFound() {
+        assertFalse(budgetDao.delete());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        JdbcTemplate template = new JdbcTemplate(dataSource);
+        template.update("DELETE FROM Budget");
```
Comment (Igor-Polatajko) [inline src/test/java/com/smartcity/dao/BudgetDaoImplTest.java]: It is better to use clearTables method from BaseTest here

File: src/main/java/com/smartcity/mapper/BudgetMapper.java [hunk 3]
```diff
@@ -2,12 +2,28 @@
 
 import com.smartcity.domain.Budget;
 import org.springframework.jdbc.core.RowMapper;
+
 import java.sql.ResultSet;
 import java.sql.SQLException;
 
 public class BudgetMapper implements RowMapper<Budget> {
 
+    private static final BudgetMapper instance = new BudgetMapper();
```
Comment (Igor-Polatajko) [inline src/main/java/com/smartcity/mapper/BudgetMapper.java]: You should add private constructor for singleton.

File: src/main/java/com/smartcity/dao/BudgetDao.java (hunk 1 above)
Comment (iholovatyi) [reply src/main/java/com/smartcity/dao/BudgetDao.java]: I think we don't have ids in the budget table. It contains only up to date value of current budget

File: src/main/java/com/smartcity/dao/BudgetDaoImpl.java [hunk 4]
```diff
@@ -1,22 +1,110 @@
...
 
+    private static final Logger logger = LoggerFactory.getLogger(BudgetDaoImpl.class);
+    private JdbcTemplate template;
+
+    @Autowired
+    public BudgetDaoImpl(DataSource source) {
+        template = new JdbcTemplate(source);
+    }
+
     public Budget create(Budget budget) {
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/BudgetDaoImpl.java]: Let's name the method:
`public Budget createOrUpdate(Budget budget)`
and remove the update and delete methods. We don't need them.

File: src/main/java/com/smartcity/dao/BudgetDaoImpl.java [hunk 5]
```diff
@@ -1,22 +1,110 @@
...
+            throw new DbOperationException(err);
+        }
+        if (amount > 0) {
+            String err = "Budget already exists in DB.";
+            logger.error(err);
+            throw new RecordExistsException(err);
+        }
+
+        try {
+            int affected = template.update(Queries.SQL_TRANSACTION_CREATE,
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/BudgetDaoImpl.java]: Do not include update logic here. Write sql query that will check if the record exists. If so, then update one, if not - then create one

File: src/main/java/com/smartcity/dao/BudgetDaoImpl.java [hunk 6]
```diff
@@ -1,22 +1,110 @@
...
+                throw new DbOperationException("Could not create Budget");
+            return budget;
+        } catch (Exception e) {
+            String err = "Failed updating Budget in DB: " + e.getMessage();
+            logger.error(err);
+            throw new DbOperationException(err);
+        }
+    }
+
+    public boolean delete() {
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/BudgetDaoImpl.java]: remove it

File: src/main/java/com/smartcity/dao/BudgetDaoImpl.java [hunk 7]
```diff
@@ -1,22 +1,110 @@
...
+            logger.error(err);
+            throw new NotFoundException(err);
+        } catch (Exception e) {
+            String err = "Failed selecting Budget from DB: " + e.getMessage();
+            logger.error(err);
+            throw new DbOperationException(err);
+        }
     }
 
     public Budget update(Budget budget) {
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/BudgetDaoImpl.java]: remove it

**Pull request merged**